daisi/src/cpps/logical/algorithms/disposition/round_robin_initiator.cpp

```diff
@@ -0,0 +1,218 @@
+// Copyright 2023 The SOLA authors
+//
+// This file is part of DAISI.
+//
+// DAISI is free software: you can redistribute it and/or modify it under the terms of the GNU
+// General Public License as published by the Free Software Foundation; version 2.
+//
+// DAISI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
+// the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+// Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along with DAISI. If not, see
+// <https://www.gnu.org/licenses/>.
+//
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include "round_robin_initiator.h"
+
+#include "cpps/amr/model/amr_fleet.h"
+
+namespace daisi::cpps::logical {
+RoundRobinInitiator::RoundRobinInitiator(std::shared_ptr<sola_ns3::SOLAWrapperNs3> sola,
+                                         std::shared_ptr<CppsLoggerNs3> logger)
+    : DispositionInitiator(sola, logger) {
+  sola::Request amr_request;
+  amr_request.all = true;
+  amr_request.permissive = true;
+  amr_request.request = "(servicetype == transport)";
+
+  amr_find_result_ = sola_->findService(amr_request);
+};
+
+bool RoundRobinInitiator::process(const AssignmentResponse &assignment_response) {
+  if (assignment_response.doesAccept()) {
+    assignment_acceptions_.push_back(assignment_response);
+    // TODO find correct task for corresponding uuid
+    // logMaterialFlowOrderStatesOfTask(assignment_response.getTaskUuid(), OrderStates::kQueued);
+  } else {
+    throw std::logic_error("the AMR should always accept the task assignment!");
+  }
+  return true;
+}
+
+// TODO
+bool RoundRobinInitiator::process(const StatusUpdate &status_update) { return true; }
+
+void RoundRobinInitiator::addMaterialFlow(std::shared_ptr<material_flow::MFDLScheduler> scheduler) {
+  if (!preparation_finished_) {
+    if (amr_find_result_.valid()) {
+      readAmrRequestFuture();
+    } else {
+      throw std::runtime_error("Future is not ready yet but required.");
+    }
+  }
+
+  material_flows_.push_back(*scheduler);
+  distributeMFTasks(material_flows_.size() - 1, false);
+}
+
+void RoundRobinInitiator::readAmrRequestFuture() {
+  for (const auto &result : amr_find_result_.get()) {
+    ParticipantInfo info;
+    float max_payload = -1;
+    std::optional<amr::LoadCarrier> loadcarrier;
```

Comment:
Why an optional? I think this is not the correct use of optional, because an optional represents that its okay to not have a value. However, if load carrier does not have a value, its an error. 

You can simply initialize the LoadCarrier with `kNoLoadCarrierType`.